Sum, Avg, Min and Max bucket pipeline aggregation

ppisljar opened the pull request:
adds bucket_sum , bucket_avg, bucket_min and bucket_max sibling pipeline aggregations

![pipeline](https://cloud.githubusercontent.com/assets/13629809/23475434/e06db314-feb7-11e6-9d1c-8901e4952273.png)

![screenshot-localhost-5601 2017-03-01 19-46-57](https://cloud.githubusercontent.com/assets/13629809/23475440/e2ea1f74-feb7-11e6-8870-9979ab4ad888.png)

Stacey-Gammon commented on src/ui/public/agg_types/metrics/bucket_avg.js:
```diff
@@ -0,0 +1,18 @@
+import { makeNestedLabel } from './lib/make_nested_label';
+import SiblingPipelineAggHelperProvider from './lib/sibling_pipeline_agg_helper';
+
+export default function AggTypeMetricDerivativeProvider(Private) {
```
These names should match how they are being imported.  E.g. `AggTypesMetricsBucketAvgProvider`.  This is why I like named exports, it forces you to use consistent naming. :)  Here and other bucket_* files

Stacey-Gammon commented on src/ui/public/agg_types/metrics/lib/sibling_pipeline_agg_controller.js:
```diff
@@ -0,0 +1,31 @@
+  return function ($scope) {
+
+    $scope.aggType = type;
+    //$scope.aggParam = $scope.agg.params[type];
```
commented out line.

Stacey-Gammon commented on src/ui/public/agg_types/metrics/bucket_avg.js:
```diff
@@ -0,0 +1,18 @@
+    name: 'avg_bucket',
+    title: 'Average Bucket',
+    makeLabel: agg => makeNestedLabel(agg, 'overall average'),
+    group: 'Pipeline Aggregations',
```
The string "Pipeline Aggregations" is repeated in many places.  What about creating a metrics_constants file to put it in?  Then if we want to change the string we only have to do it in one spot.

Stacey-Gammon commented on src/ui/public/agg_types/metrics/lib/sibling_pipeline_agg_controller.js:
```diff
@@ -0,0 +1,31 @@
+    $scope.aggType = type;
+    //$scope.aggParam = $scope.agg.params[type];
+    $scope.aggTitle = type === 'customMetric' ? 'Metric' : 'Bucket';
+    $scope.aggGroup = type === 'customMetric' ? 'metrics' : 'buckets';
```
Looks like there is `aggGroup` and also `agg.group` and they mean different things.  I'm not sure I have a suggestion on the naming, but it'd be nice if they could be distinguished better.

Almost seems like maybe typeName would be more fitting?  I think I tracked this down to being used here:
`aggTypes.byType[$scope.groupName];`

Though that might be too far reaching to change it in this PR since that name already exists and group is the new name.

I think something else that would be worthwhile as a separate PR would be moving `metrics` and `buckets` to constants too.  That would be easier to understand the importance of these strings and where else they are being referenced.  Then you could do something like:
```
export const AggTypes = {
  METRICS: 'metrics',
  BUCKETS: 'buckets'
}
```

Food for thought mostly, I understand if you prefer to leave as is for lack of a better alternative at this point.

ppisljar replied:
i renamed agg.group to agg.subtype

Stacey-Gammon commented on src/ui/public/agg_types/metrics/lib/sibling_pipeline_agg_controller.js:
```diff
@@ -0,0 +1,31 @@
+    $scope.aggGroup = type === 'customMetric' ? 'metrics' : 'buckets';
+    $scope.safeMakeLabel = safeMakeLabel;
+
+    $scope.rejectAgg = function (agg) {
```
Is this used at all?  I'm a little confused at the interaction and differences between rejectAgg and the aggFilter.  Since the aggFilter already contains all these, if I select for example `Top Hits`, I don't see any of the sibling  drop downs.

ppisljar replied:
you are right, this is not used here ... to much copy pasting from parent pipeline agg :)

Stacey-Gammon commented:
Ran into an error using Sum Bucket and Terms agg:

<img width="1271" alt="screen shot 2017-02-15 at 11 12 41 am" src="https://cloud.githubusercontent.com/assets/16563603/22983323/bb4dc43c-f36f-11e6-9d86-17d873d69618.png">

ppisljar commented:
@Stacey-Gammon the reason for error was sorting on bucket metric in your terms bucket. i updated it so its not allowed anymore. i also addressed your other comments. thanks a lot for your super fast response !

tbragin commented:
@ppisljar A couple of quick comments:
* I do love the dropdown calling out aggregation types, nice call @alexfrancoeur! However, instead of "Basic Aggregations" I think we should call them what [ES documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics.html) calls them, which is "Metrics Aggregations" 
<img width="400" alt="screen shot 2017-02-20 at 4 43 21 pm" src="https://cloud.githubusercontent.com/assets/933397/23146453/c1e4749c-f78b-11e6-8d49-b405a5f45da7.png">

* I also think we should order bucket aggregations the same as metrics aggregations, with "Average" preceding "Sum". 
* It'd be nice if, once fully configured, switching between Sum, Average, Min, and Max Bucket Aggregations didn't wipe out the configuration below it. You'll see that https://github.com/elastic/kibana/pull/9725 allows to switch between all these bucket aggs w/o user having to rebuild their config.
* I got this error when selecting the following option
<img width="1439" alt="screen shot 2017-02-20 at 5 56 08 pm" src="https://cloud.githubusercontent.com/assets/933397/23148012/19bb9d56-f798-11e6-959e-ff1456a09c89.png">

* You kind of do have to know what you're doing with Bucket aggregations in the context of Date Histogram -- it's certainly not for the faint-hearted. I think I finally figured out how to configure something SLA-like "bucket_average(max(bytes,1h),24h), split by OS" (see screenshots). Anything we can do to reduce the vertical amount of space needed for these configuration would improve the experience, imo. For one, I think we should prioritize removing unnecessary custom labels: https://github.com/elastic/kibana/issues/10217 (I would go so far to even say that it should be a blocker for 5.4 if we plan to ship pipeline aggs as part of that release). Secondarily, I wonder if additional levels of collapsing would be useful, though I'd imagine it's a harder endeavor.
<img width="1440" alt="screen shot 2017-02-20 at 6 10 38 pm" src="https://cloud.githubusercontent.com/assets/933397/23148097/7b51211c-f798-11e6-9a95-7cefe3fd39a8.png">
<img width="1440" alt="screen shot 2017-02-20 at 6 10 49 pm" src="https://cloud.githubusercontent.com/assets/933397/23148098/7b5204a6-f798-11e6-983e-47cf1619b4a4.png">

* Metric configuration is less staggering, since you can't split by anything, but superflous labels make it look convoluted as well. 
<img width="1440" alt="screen shot 2017-02-20 at 7 09 23 pm" src="https://cloud.githubusercontent.com/assets/933397/23149376/5c8f729e-f7a0-11e6-8a9d-97a462082522.png">

tbragin commented:
@ppisljar A couple of other problems I ran into... Do we allow pipeline agg of a pipeline or should that be disabled in the vis editor?

<img width="1440" alt="screen shot 2017-02-21 at 9 36 38 am" src="https://cloud.githubusercontent.com/assets/933397/23176796/74bb8410-f819-11e6-8226-5c9b77de3125.png">
<img width="1440" alt="screen shot 2017-02-21 at 9 37 20 am" src="https://cloud.githubusercontent.com/assets/933397/23176797/74bf5838-f819-11e6-82dd-fd6ccf0e64aa.png">

ppisljar commented:
- sorting on buckets is not allowed (in terms)
- buckets on derivative are not allowed
- derivative on buckets IS allowed

- removing the labels: i agree, but not sure if i would make it a blocker ....
- switching without rebuilding: this is gonna be hard at this point (for 5.4) unless we want to go forward with a hacky solution (i oppose). there is no mechanism in place at the moment to allow this. If you are ok with tracking this separately please open an issue and link it here.

thomasneirynck commented on src/ui/public/agg_types/buckets/terms.js:
```diff
@@ -16,7 +16,12 @@ export default function TermsAggDefinition(Private) {
   const routeBasedNotifier = Private(routeBasedNotifierProvider);
 
-  const aggFilter = ['!top_hits', '!percentiles', '!median', '!std_dev', '!derivative'];
+  const aggFilter = [
```
resolve merge conflict

thomasneirynck commented:
Could we split 'pipeline aggregations' up between parent and sibling aggregations? ie. have 3 sections iso 2.

![image](https://cloud.githubusercontent.com/assets/1833023/23387545/42d965fe-fd2b-11e6-9469-2e2fc505235e.png)

The reason for this is that the metric configuration is different for each. It is a little confusing to see the UI form change.

ppisljar commented:
thanks @thomasneirynck 
- i split aggs in 3 sections
- rebased on master

- regarding custom metric dropdown ... it will have multiple values if you have more metrics. besides, thats out of scope of this pr (bucket pipeline aggs don't have the metrics dropdown)

ppisljar commented:
jenkins, test this

Stacey-Gammon commented:
My choices here don't seem to be updating the y axis at all.  Is this a bug or am I doing something wrong?  I'm trying to create a visualization like the person in that discuss ticket yesterday was hoping to create.

![bucketshisto](https://cloud.githubusercontent.com/assets/16563603/23411408/32cb5148-fda0-11e6-92ad-a0d9da18f27a.gif)

thomasneirynck commented:
@Stacey-Gammon That's because of the nature of the bucket sum, in that particular configuration. It does an overall sum of the counts, which is identical for all those fields (given that the same docs are included in all those buckets-by-date).

choose a metric other than count, and then you will see different results (e.g. min of bytes, ...)

Stacey-Gammon commented on src/core_plugins/metrics/public/components/__tests__/add_delete_buttons.js:
```diff
@@ -0,0 +1,67 @@
+// import React from 'react';
```
this sneak in by mistake? :)

ppisljar commented:
future improvements: #10633

epixa commented:
@ppisljar Can you add a detailed description (with screenshots) of these features to this PR description?  Whenever we make any progress on pipeline aggs, please also drop a comment on the pipeline agg ticket since *a lot* of people want to know about this.
